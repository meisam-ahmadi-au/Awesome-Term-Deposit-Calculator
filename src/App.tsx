import React, { useState } from "react";
import InterestPaidSelectField from "./components/interest-paid-select-field/interest-paid-select-field";
import Button from "./components/button/button.component";
import NumberInputField from "./components/number-input-fielld/number-input-field.component";
import { calculateFinalBalance } from "./utils/calculate-term-deposit-balance";

const App: React.FC = () => {
  const [depositAmount, setDepositAmount] = useState<number>();
  const [interestRate, setInterestRate] = useState<number>();
  const [investmentTerm, setInvestmentTerm] = useState<number>();
  const [interestPaid, setInterestPaid] = useState<number>(1);
  const [finalBalance, setFinalBalance] = useState<string>();
  const isCalculateButtonDisabled = !(
    depositAmount &&
    interestRate &&
    investmentTerm &&
    interestPaid
  );

  const calculateHandler = () => {
    if (depositAmount && interestRate && investmentTerm) {
      const amount = calculateFinalBalance(
        depositAmount,
        interestRate,
        investmentTerm,
        interestPaid
      );
      setFinalBalance(amount?.toFixed(2));
    }
  };

  return (
    <div className="max-w-md mx-auto p-6">
      <h1 className="text-xl font-bold mb-4">Term Deposit Calculator</h1>
      <NumberInputField
        label="Start Deposit Amount"
        setter={setDepositAmount}
        min={1000}
        max={1500000}
      />
      <NumberInputField
        label="Interest Rate (%)"
        setter={setInterestRate}
        min={0.1}
        max={15}
      />
      <NumberInputField
        label="Investment Term (years)"
        setter={setInvestmentTerm}
        min={1}
        max={15}
      />

      <InterestPaidSelectField setter={setInterestPaid} />

      <Button
        label="Calculate"
        onClick={calculateHandler}
        disabled={isCalculateButtonDisabled}
      />

      {finalBalance !== null && (
        <div className="mt-4">
          <h2 className="text-lg font-semibold">
            Final Balance: ${finalBalance}
          </h2>
        </div>
      )}
    </div>
  );
};

export default App;
