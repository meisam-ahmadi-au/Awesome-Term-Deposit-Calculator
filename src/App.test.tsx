import { render, screen, fireEvent } from "@testing-library/react";
import App from "./App";

describe("App Component", () => {
  test("matches the snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  test("renders initial UI correctly and disables calculate button", () => {
    render(<App />);

    expect(screen.getByLabelText(/start deposit amount/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/interest rate/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/investment term/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/interest paid/i)).toBeInTheDocument();
    const calculateButton = screen.getByRole("button", { name: /calculate/i });
    expect(calculateButton).toBeDisabled();
  });

  test("enables the calculate button when valid inputs are provided", () => {
    render(<App />);

    const depositAmountInput = screen.getByLabelText(/start deposit amount/i);
    const interestRateInput = screen.getByLabelText(/interest rate/i);
    const investmentTermInput = screen.getByLabelText(/investment term/i);

    fireEvent.change(depositAmountInput, { target: { value: 10000 } });
    fireEvent.change(interestRateInput, { target: { value: 1.5 } });
    fireEvent.change(investmentTermInput, { target: { value: 3 } });

    const calculateButton = screen.getByRole("button", { name: /calculate/i });
    expect(calculateButton).not.toBeDisabled();
    fireEvent.click(calculateButton);
    const finalBalance = screen.getByText(/final balance/i);
    expect(finalBalance).toBeInTheDocument();
    expect(finalBalance).toHaveTextContent(/\$10456/);
  });

  test("show feedback to user", () => {
    render(<App />);
    const depositAmountInput = screen.getByLabelText(/start deposit amount/i);
    fireEvent.change(depositAmountInput, { target: { value: "1222w" } });
    const errorMessage = screen.getByText(/Please enter a valid number/i);
    expect(errorMessage).toBeInTheDocument();
  });
});
