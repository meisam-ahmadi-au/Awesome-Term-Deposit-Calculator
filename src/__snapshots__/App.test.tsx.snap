// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > matches the snapshot 1`] = `
<div>
  <div
    class="max-w-md mx-auto p-6"
  >
    <h1
      class="text-xl font-bold mb-4"
    >
      Term Deposit Calculator
    </h1>
    <div
      class="mb-4 "
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
      >
        Start Deposit Amount
        <input
          class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500 sm:text-sm"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="mb-4 "
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
      >
        Interest Rate (%)
        <input
          class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500 sm:text-sm"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="mb-4 "
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
      >
        Investment Term (years)
        <input
          class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500 sm:text-sm"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="mb-4 "
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
      >
        Interest Paid
        <select
          class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500 sm:text-sm"
        >
          <option
            value="monthly"
          >
            Monthly
          </option>
          <option
            value="quarterly"
          >
            Quarterly
          </option>
          <option
            value="annually"
          >
            Annually
          </option>
          <option
            value="maturity"
          >
            At Maturity
          </option>
        </select>
      </label>
    </div>
    <div
      class="mb-4 "
    >
      <button
        class="w-full px-4 py-2 rounded-md focus:outline-none bg-gray-400 text-gray-200 cursor-not-allowed"
        disabled=""
      >
        Calculate
      </button>
    </div>
    <div
      class="mt-4"
    >
      <h2
        class="text-lg font-semibold"
      >
        Final Balance: $
      </h2>
    </div>
  </div>
</div>
`;
